src/polydatum/middleware.py

```diff
@@ -0,0 +1,184 @@
+from typing import Tuple, Any, Dict, Optional, Callable
+
+from polydatum.services import Service
+from polydatum.context import DataAccessContext
+
+
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name = None
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
+        return all([
```

Comment:
`all` is not needed here. It actually introduces a bug.

3 parts will be evaluated even If one of the other parts is false. Execution will not stop early. So `self` may not be a `PathSegment`, but `self.name == other.name` will still evaluate. 

In this case, it's impossible for `isinstance(self, PathSegment)` to not be a `PathSegment` because this is on the `PathSegment` class. I think you meant `isinstance(other, PathSegment)`.

This is the best way to do it:

```python
return isinstance(other, PathSegment) and self.name == other.name and self.meta == other.meta
```

1. Execution will stop if any of the left sides return false, saving 0.00000000000001 bitcoins worth of electricity
2. No list reservation needed